clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -3313,6 +3313,11 @@ static void handleSYCLIntelFPGADisableLoopPipeliningAttr(Sema &S, Decl *D,
   if (checkAttrMutualExclusion<SYCLIntelFPGAInitiationIntervalAttr>(S, D, A))
     return;
 
+  // [[intel::disable_loop_pipelining] and [[intel::max_concurrency()]]
+  // attributes are incompatible.
+  if (checkAttrMutualExclusion<SYCLIntelFPGAMaxConcurrencyAttr>(S, D, A))
+    return;
+
```

Comment:
Happy days are here, we don't have to do this sort of dance any longer! You can remove this code (and some below) and instead do this all from Attr.td.

Below whichever attribute is defined last between disable_loop_pipelining and max_concurrency, add a definition to Attr.td like:
```
def : MutualExclusions<[SYCLIntelFPGAMaxConcurrency, SYCLIntelFPGADisableLoopPipelining]>;
```
Tablegen will then provide the checking logic for you so you don't have to explicitly add code for it. The only time we need to use `checkAttrMutualExclusion()` for declaration attributes is when the mutual exclusion has other logic to it (like ignoring implicitly-added attributes).

Reply:
This would not work with the patch: 

https://github.com/intel/llvm/pull/3507/

Reply:
Oh, I see -- it's because of the `DeclOrStmtAttr` base class, that makes sense. Thanks!

Reply:
> Oh, I see -- it's because of the `DeclOrStmtAttr` base class, that makes sense. Thanks!

Yes.

Reply:
> I will hold this PR now until #3507 gets merged.

Yes, please, I had one (possibly minor) question there and it should get approved soon.

Reply:
Done